File: backend_deno_refactor/domain/schedule/event/birthday/create.ts
import type { Result } from "../../../lang/result.ts";
import type { IdGenerator } from "../../../generator.ts";
import type { DateGenerator } from "../../../generator.ts";
import type { RepoErr } from "../../../repo.ts";
import type { ValidationErr } from "../../../validation/validate.ts";
import type { Schema } from "../../../validation/schema.ts";
import type { User } from "../../user/model.ts";
import type { EventRepo } from "../repo.ts";
import type { EventCreate } from "../create.ts";
import type { Event } from "../model.ts";
import { validateSchema } from "../../../validation/validate.ts";
import { eventCreate } from "../create.ts";

export type BirthdayCreate = {
    readonly name: Event["name"];
    readonly day: Event["day"];
};

const birthdayCreateSchema: Schema<BirthdayCreate> = {
    name: [
        { type: "str" },
        { type: "strMinLen", min: 1 },
        { type: "strMaxLen", max: 32 },
    ],
    day: [
        { type: "dt" },
        { type: "dtMin", min: "1970-01-01" },
    ],
};

export function birthdayCreateToEventCreate(event: BirthdayCreate): EventCreate {
    return {
        name: event.name,
        day: event.day,
        begin: "00:00",
        end: "23:59",
        category: "BIRTHDAY",
        frequency: "1Y",
        weekendRepeat: false,
    };
}

type BirthdayCreateErrors =
    | RepoErr
    | ValidationErr;

export function birthdayCreateService(
    repo: EventRepo,
    idGenerator: IdGenerator,
    dateGenerator: DateGenerator,
    userId: User["id"],
    event: BirthdayCreate,
): Promise<Result<Event, BirthdayCreateErrors>> {
    const schemaValidation = validateSchema(birthdayCreateSchema, event);
    if (schemaValidation.type === "err") {
        return Promise.resolve(schemaValidation);
    }
    const builtEvent = birthdayCreateToEventCreate(event);
    return eventCreate(repo, idGenerator, dateGenerator, builtEvent, userId);
}

File: backend/src/domain/schedule/event/appointment/create.rs
use std::{collections::HashMap, sync::LazyLock};

use crate::domain::{generator::{DateTimeGen, IdGen}, schedule::event::{
    create::{event_c, EventC}, error::EventErr, model::{Event, EventCat, EventFreq}, repo::EventRepo
}, validation::{Schema, Validator, Value, V}};

pub struct AppointmentC {
    pub name: String,
    pub day: String,
    pub begin: String,
    pub end: String,
    pub frequency: Option<String>,
    pub weekend_repeat: Option<bool>,
}

static APPOINTMENT_C_SCHEMA: LazyLock<Schema> = LazyLock::new(|| {
    HashMap::from([
        ("name", vec![V::Required, V::Str, V::StrMinLen(1), V::StrMaxLen(32)]),
        ("day", vec![V::Required, V::Dt, V::DtMin(String::from("1970-01-01"))]),
        ("begin", vec![V::Required, /*V::Time*/]),
        ("end", vec![V::Required, /*V::Time, V::Gt("begin")*/]),
        ("frequency", vec![/* V::Enum(vec!["1D", "2D", "1W", "1M", "3M", "6M", "1Y", "2Y"]) */]),
        ("weekend_repeat", vec![V::Bool])
    ])
});

pub fn event_c_from_appointment_c(event: AppointmentC) -> EventC {
    EventC {
        name: event.name,
        begin: event.begin,
        end: event.end,
        category: EventCat::Appointment,
        frequency: event.frequency.and_then(|freq| EventFreq::parse(&freq)),
        weekend_repeat: event.weekend_repeat,
    }
}

pub fn event_appointment_c(
    validator: &dyn Validator,
    repo: &dyn EventRepo,
    id_gen: &dyn IdGen,
    date_time_gen: &dyn DateTimeGen,
    appointment_c_model: AppointmentC,
    user_id: String,
) -> Result<Event, EventErr> {
    let input_value = Value::Obj(HashMap::from([
        (String::from("name"), Value::Str(appointment_c_model.name.clone())),
        (String::from("day"), Value::Str(appointment_c_model.day.clone())),
        (String::from("begin"), Value::Str(appointment_c_model.begin.clone())),
        (String::from("end"), Value::Str(appointment_c_model.end.clone())),
        (String::from("frequency"), match appointment_c_model.frequency.clone() {
            Some(v) => Value::Str(v),
            None => Value::Absent,
        }),
        (String::from("weekend_repeat"), match appointment_c_model.weekend_repeat.clone() {
            Some(v) => Value::Bool(v),
            None => Value::Absent,
        }),
    ]));
    validator.validate(&APPOINTMENT_C_SCHEMA, &input_value).map_err(EventErr::Schema)?;
    let event_c_model = event_c_from_appointment_c(appointment_c_model);
    return event_c(repo, id_gen, date_time_gen, event_c_model, user_id);
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_event_c_from_appointment() {
        let appointment = AppointmentC {
            name: String::from("Dentist"),
            day: String::from("2024-03-31"),
            begin: String::from("18:00"),
            end: String::from("22:00"),
            frequency: Some(String::from("2D")),
            weekend_repeat: Some(true),
        };
        let create_event = EventC {
            name: String::from("Dentist"),
            begin: String::from("18:00"),
            end: String::from("22:00"),
            category: EventCat::Appointment,
            frequency: Some(EventFreq::D2),
            weekend_repeat: Some(true),
        };
        assert_eq!(event_c_from_appointment_c(appointment), create_event);
    }
}
